packages/application/src/requests.tsx

```diff
@@ -0,0 +1,137 @@
+    const timestamp = response.timestamp ? response.timestamp : '';
+    const traceback = response.traceback ? response.traceback : '';
+    const default_body = response.timestamp
+      ? 'Check the JupyterLab log for more details at ' + response.timestamp
+      : 'Check the JupyterLab log for more details';
+
+    return showDialog({
+      title: 'Error making request',
+      body:
+        reason || message ? (
+          <ErrorDialogContent
+            reason={reason}
+            message={message}
+            timestamp={timestamp}
+            traceback={traceback}
+            default_msg={default_body}
+          />
+        ) : (
+          <p>{default_body}</p>
+        ),
+      buttons: [Dialog.okButton()]
+    });
+  }
+
+  static server404(): Promise<Dialog.IResult<any>> {
+    return showDialog({
+      title: 'Error contacting server',
+      body: <p>Elyra service endpoint not found.</p>,
+      buttons: [Dialog.okButton()]
+    });
+  }
+
+  static async makeGetRequest(
+    requestExt: string,
+    displayWaitDialog: boolean
+  ): Promise<any> {
+    return this.makeServerRequest(
+      requestExt,
+      { method: 'GET' },
+      displayWaitDialog
+    );
+  }
+
+  static async makePostRequest(
+    requestExt: string,
+    requestBody: any,
+    displayWaitDialog: boolean
+  ): Promise<any> {
+    return this.makeServerRequest(
+      requestExt,
+      { method: 'POST', body: requestBody },
+      displayWaitDialog
+    );
+  }
+
+  static async makeServerRequest(
+    requestExt: string,
+    requestOptions: any,
+    displayWaitDialog: boolean
+  ): Promise<any> {
+    // use ServerConnection utility to make calls to Jupyter Based services
+    // which in this case are the in the extension installed by this package
+    const settings = ServerConnection.makeSettings();
+    const requestUrl = URLExt.join(settings.baseUrl, requestExt);
+
+    console.log(
+      'Sending a ' + requestOptions.method + ' request to ' + requestUrl
+    );
+
+    const waitDialog: Dialog<any> = new Dialog({
+      title: 'Making server request...',
+      body: 'This may take some time',
+      buttons: [Dialog.okButton()]
+    });
+
+    if (displayWaitDialog) {
+      waitDialog.launch();
+    }
+
+    const getServerResponse: Promise<any> = new Promise((resolve, reject) => {
+      ServerConnection.makeRequest(requestUrl, requestOptions, settings).then(
+        (response: any) => {
+          if (displayWaitDialog) {
+            waitDialog.resolve();
+          }
+
+          response.json().then(
+            (result: any) => {
+              if (!HTML_STATUS_SUCCESS.includes(response.status)) {
+                return this.serverError(result);
+              }
+
+              resolve(result);
+            },
+            (reason: any) => {
+              // handle 404 if elyra server extension is not found
+              return this.server404();
+            }
+          );
+        }
+      );
+    });
```

Comment:
I find the flow here confusing, and there might be exceptions not being caught, then serverResponse could never be resolved. I believe at the end of every possible end point should be either a resolve or reject, instead of returns.
- ServerConnection.makeRequest doesn't have a catch for promise rejection
- if response.json fails, does it fall into this.server404? 
- if response.json gets a result with a server error status, shouldn't line 120 resolve the promise instead of using return?

Reply:
@karlaspuldaro and I discussed this in detail over Webex to get both our head around how it work and I will be adding in-line documentation to explain it, no code changes are required.

Reply:
added comments

Reply:
I still think this is an important piece since it's being called by any get and post requests, and as agreed in our discussion, it at least requires more testing for edge cases to make sure we don't have problems maintaining this in the future. 

I tried running  the prototype below in node to see what happens when something goes wrong somewhere inside, await doesn't seem to catch all exceptions if they're not properly handled:
```
const asyncFunc = async () => {
	// First promise
	const p1 = new Promise((res, rej) => {
		
		// Second promise
		new Promise((res2,  rej2) => { 
		setTimeout(
			() => {
				// Let's say there is some sort of error in this inner promise
				console.log("rejecting second promise");
				rej2("exception: rej2");
			}
			, 1000)}
		)
		.then(
			result => {
				if(!2===2){
					return console.log("some server error");
					// why not resolve
				}
				resolve(result);
			},
			error =>  {
				// Exception catch block
				console.log("404");
				// do we want to resolve or reject here?
			}
		);

	}); // no exception catch block for p1
	

	const result = await p1;
	console.log("p1 done! p1 res = "  + result + "\nReturning...");
	return result;
}


asyncFunc();
console.log("waiting for promise...");
```
And my output was:

```
waiting for promise...
rejecting second promise
404
```
In this case, makeServerRequest never gets resolved and its return is never called, even if it called 404 callback and the error bubbled up.
Even if we only treat the specific cases we want our flow to fall into and not expose anything unnecessary to the client, maybe we should make sure every possible promise path is being handled.
I'm not totally sure what are the consequences of hanging promises, or even if our flow is somehow controlled and changes are not required, test scenarios can answer this question.

Reply:
ok so your argument is that we should "handle" all the possible error cases, even if we should reach them. (I would just raise an error of some sort in those cases, not a dialog)